transfers/esgf_holdings.py

```diff
@@ -0,0 +1,197 @@
+
+
+def make_model_variants_lut(tc, node_ep, node_prefix, models, scenarios, ncpus):
+    """Create a lookup table of all variants available for each model/ scenario combination. Uses an existing TransferClient object.
+
+    Returns a pandas DataFrame with a list of variants available for each model/scenario
+    """
+    # find what variants are available for each model / scenario combination
+    args = list(product(["CMIP"], models, ["historical"])) + list(
+        product(["ScenarioMIP"], models, scenarios)
+    )
+    # include the TransferClient object for each query
+    args = [[tc, node_ep, node_prefix] + list(a) for a in args]
+
+    with Pool(ncpus) as pool:
+        rows = pool.starmap(list_variants, args)
+
+    df = pd.DataFrame(rows)
+
+    return df.dropna()
+
+
+def get_filenames(
+    tc, node_ep, node_prefix, activity, model, scenario, variant, frequency, varname
+):
+    """Get the file names for a some combination of model, scenario, and variable."""
+    # the subdirectory under the variable name is the grid type.
+    #  This is almost always "gn", meaning the model's native grid, but it could be different.
+    #  So we have to check it instead of assuming. I have only seen one model where this is different (gr1, GFDL-ESM4)
+    var_path = node_prefix.joinpath(
+        activity,
+        model_inst_lu[model],
+        model,
+        scenario,
+        variant,
+        frequency,
+        varname,
+    )
+    grid_type = utils.operation_ls(tc, node_ep, var_path)
+    empty_row = {
+        "model": model,
+        "scenario": scenario,
+        "variant": variant,
+        "frequency": frequency,
+        "variable": varname,
+        "grid_type": None,
+        "version": None,
+        "n_files": None,
+        "filenames": None,
+    }
+    if isinstance(grid_type, int):
+        # error if int
+        # there is no data for this particular combination.
+        row_di = empty_row
+    else:
+        # combo does exist, return all filenames
+        grid_type = grid_type[0]
+        versions = utils.operation_ls(tc, node_ep, var_path.joinpath(grid_type))
+
+        if len(versions) == 0:
+            # in rare cases, the grid_type folder exists but with no data.
+            row_di = empty_row
+            # setting grid_type value will allow us to know where this occurs.
+            row_di["grid_type"] = grid_type
+        else:
+            use_version = sorted(versions)[-1]
+            fns = utils.operation_ls(
+                tc, node_ep, var_path.joinpath(grid_type, use_version)
+            )
+            row_di = {
+                "model": model,
+                "scenario": scenario,
+                "variant": variant,
+                "frequency": frequency,
+                "variable": varname,
+                "grid_type": grid_type,
+                "version": use_version,
+                "n_files": len(fns),
+                "filenames": fns,
+            }
```

Comment:
huh, thanks for catching this Craig and for digging in a bit, I will look into this next week and test that potential fix. Thanks for the great review and suggestions!

Reply:
Well I tried running it again and did not get any 404's, so I'm kinda puzzled as to why you would have! I don't know what causes a 404. From what I can tell, if `use_version` gets set in that snippet we're looking at, then it seems like the path resolved by `var_path.joinpath(grid_type, use_version)` _must_ exist, and so then why would it 404? Maybe it has something to do with calling the `utils.operation_ls` in parallel. 

Would you mind re-running this script with something like this in place of the snippet above, so that I can have a look at the CSV produced and try to dig in further? (Minor one-line difference, just adding the "404" in as a way to differentiate from the empty rows caused by the empty `grid_type` paths) 

```
        if versions == 404 or len(versions) == 0:
            # in rare cases, the grid_type folder exists but with no data.
            row_di = empty_row
            # setting grid_type value will allow us to know where this occurs.
            row_di["grid_type"] = grid_type
        else:
            use_version = sorted(versions)[-1]
            fns = utils.operation_ls(
                tc, node_ep, var_path.joinpath(grid_type, use_version)
            )
            if fns == 404:
                row_di = empty_row
                row_di.update("version": "404")
            else:
                row_di = {
                    "model": model,
                    "scenario": scenario,
                    "variant": variant,
                    "frequency": frequency,
                    "variable": varname,
                    "grid_type": grid_type,
                    "version": use_version,
                    "n_files": len(fns),
                    "filenames": fns,
                }
```

Reply:
I ran `python esgf_holdings.py --node llnl` several more times using the following chunk of code to detect any type of HTTP error (and time `utils.operation_ls` returns an integer):

```
    if isinstance(grid_type, int):
        # error if int
        # there is no data for this particular combination.
        row_di = empty_row
    else:
        # combo does exist, return all filenames
        grid_type = grid_type[0]
        versions = utils.operation_ls(tc, node_ep, var_path.joinpath(grid_type))

        if isinstance(versions, int) or len(versions) == 0:
            # in rare cases, the grid_type folder exists but with no data.
            row_di = empty_row
            # setting grid_type value will allow us to know where this occurs.
            row_di["grid_type"] = grid_type
            row_di["version"] = versions
        else:
            use_version = sorted(versions)[-1]
            fns = utils.operation_ls(
                tc, node_ep, var_path.joinpath(grid_type, use_version)
            )
            if isinstance(fns, int):
                row_di = empty_row
                row_di["version"] = fns
            else:
                row_di = {
                    "model": model,
                    "scenario": scenario,
                    "variant": variant,
                    "frequency": frequency,
                    "variable": varname,
                    "grid_type": grid_type,
                    "version": use_version,
                    "n_files": len(fns),
                    "filenames": fns,
                }
```

I no longer received 404 errors, but started getting intermittent HTTP 502 errors that occurred for a (seemingly) random set of paths for each run. My final run succeeded without any HTTP errors, though, so I'm marking this as approved for now and we can keep an eye on this in the future.